flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.calcite.schema.Table;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static java.lang.String.format;
+
+/**
+ * A mapping between Flink catalog's database and Calcite's schema.
+ * Tables are registered as tables in the schema.
+ */
+class DatabaseCalciteSchema implements Schema {
+	private final String databaseName;
+	private final String catalogName;
+	private final Catalog catalog;
+
+	public DatabaseCalciteSchema(String databaseName, String catalogName, Catalog catalog) {
+		this.databaseName = databaseName;
+		this.catalogName = catalogName;
+		this.catalog = catalog;
+	}
+
+	@Override
+	public Table getTable(String tableName) {
+		ObjectPath tablePath = new ObjectPath(databaseName, tableName);
+
+		try {
+			if (!catalog.tableExists(tablePath)) {
+				return null;
+			}
+
+			CatalogBaseTable table = catalog.getTable(tablePath);
+
+			// TODO supports GenericCatalogView
+			if (table instanceof QueryOperationCatalogView) {
+				QueryOperationCatalogView view = (QueryOperationCatalogView) table;
+				QueryOperation operation = view.getQueryOperation();
+				if (operation instanceof DataStreamQueryOperation) {
+					List<String> qualifiedName = Arrays.asList(catalogName, databaseName, tableName);
+					((DataStreamQueryOperation) operation).setQualifiedName(qualifiedName);
+				}
+				return QueryOperationCatalogViewTable.createCalciteTable(view);
+			} else if (table instanceof ConnectorCatalogTable) {
+				ConnectorCatalogTable<?, ?> connectorTable = (ConnectorCatalogTable<?, ?>) table;
+				return connectorTable.getTableSource()
+					.map(tableSource -> new TableSourceTable<>(
+						tableSource,
+						!connectorTable.isBatch(),
+						FlinkStatistic.UNKNOWN())
+					).orElseThrow(() -> new TableException("Cannot query a sink only table."));
+			} else if (table instanceof FlinkTempCatalogTable) {
+				return ((FlinkTempCatalogTable) table).getAbstractTable();
+			} else {
+				throw new TableException("Unsupported table type: " + table);
+			}
+		} catch (TableNotExistException | CatalogException e) {
+			// TableNotExistException should never happen, because we are checking it exists
+			// via catalog.tableExists
+			throw new TableException(format(
+				"A failure occurred when accessing table. Table path [%s, %s, %s]",
+				catalogName,
+				databaseName,
+				tableName), e);
+		}
+	}
+
+	@Override
+	public Set<String> getTableNames() {
+		try {
+			return new HashSet<>(catalog.listTables(databaseName));
+		} catch (DatabaseNotExistException e) {
+			throw new CatalogException(e);
+		}
+	}
+
+	@Override
+	public RelProtoDataType getType(String name) {
+		return null;
+	}
+
+	@Override
+	public Set<String> getTypeNames() {
+		return new HashSet<>();
+	}
+
+	@Override
+	public Collection<Function> getFunctions(String s) {
+		return new HashSet<>();
+	}
+
+	@Override
+	public Set<String> getFunctionNames() {
```

Comment:
Some of the interfaces inherited from Calcite `Schema` seems useless now, can we have a uniform impl for all these Flink schema implementations ?

Reply:
make sense